docs/source/concepts/versioning.md

```diff
@@ -0,0 +1,104 @@
+
+### What can I do with Hub versioning currently?
+* Modify dataset elements across different versions
+* Seamlessly switch between versions
+
+### Features coming in the future
+* Modify schema across versions (add or remove Tensors)
+* Track versions across transforms
+* Delete branches
+* (More suggestions welcome!)
+
+## Core Concepts
+
+### Saving the state of a dataset
+The current state of a dataset can be saved using commit which is very similar to git commit. commit takes a message as a parameter, saves the current state of the dataset and returns the commit id generated. You can use this commit id to come back to this commit at any time. You don’t necessarily need to store this commit id, it can be accessed later using log (similar to git log).
+Example:-
+```python
+ds = hub.Dataset("path/to/dataset")
+for i in range(10):
+    ds["img", i] = i * np.ones((100, 100, 3))  # assuming the dataset has a tensor called img
+a = ds.commit("my first commit") # the commit message is optional, but recommended
+```
+
+### Changing branches
+Similar to git, you start out on the master branch by default. There are 2 ways to switch to another branch:-
+* Using checkout and passing the branch name as the address
+* Making changes while not on the head of a branch i.e. in a state similar to the “Detached head" state in git. For instance, if you have made 5 commits on master and go back to the second commit (again using checkout, explained in detail in the next section) and try to make some change there, you will automatically get checked out to a new branch with a name similar to “auto:ec33fced9d75021a32ae28ff”. Alternatively you could also manually checkout to a branch from this state yourself and prevent auto checkout.
+
+```python
+ds = hub.Dataset("path/to/dataset")
+ds.checkout("alternate", create=True)  # creates a new branch
+
+ds.checkout("master")
+for i in range(10):
+    ds["img", i] = 2 * i * np.ones((100, 100, 3))
+a = ds.commit("first commit")
+for i in range(10):
+    ds["img", i] = 3 * ds["img", i].compute()
+b = ds.commit("second commit")
+ds.checkout(a)
+ds["img", 6] = 4 * ds["img", i].compute()  # auto checks out to new branch
+c = ds.commit("third commit")
+```
+
+
+### Switching between different versions of the dataset
+Checkout is used to switch between different versions of the dataset. It takes a commit id or a branch name as a parameter and changes the dataset state to that. It can also be used to create a new branch from the current version by passing the create parameter as True.
+When checking out from the head of a branch, the current uncommitted changes (if any) are also brought over to the new branch.
+Example:-
+```python
+ds = hub.Dataset("path/to/dataset")
+ds.checkout("alternate", create=True)  # creates a new branch
```

Comment:
Have we reimplemented the feedback regarding create parameter?

Reply:
From the slack discussion:-
"I kept create=True in checkout to prevent accidental checkout to a new branch in case of a typo."
I'm not sure if we want to remove that. I think it's the same reason why git uses checkout -b to signify new branch.